File: linux/ui-tauri/src/lib/sasEmoji.ts

export interface SasGlyph {
  emoji: string;
  name: string;
}

export const SAS_EMOJI: SasGlyph[] = [
  { emoji: "🦊", name: "Fox" },        { emoji: "🐼", name: "Panda" },
  { emoji: "🦁", name: "Lion" },       { emoji: "🐯", name: "Tiger" },
  { emoji: "🐶", name: "Dog" },        { emoji: "🐱", name: "Cat" },
  { emoji: "🐵", name: "Monkey" },     { emoji: "🐸", name: "Frog" },
  { emoji: "🐧", name: "Penguin" },    { emoji: "🦉", name: "Owl" },
  { emoji: "🦅", name: "Eagle" },      { emoji: "🐝", name: "Bee" },
  { emoji: "🦋", name: "Butterfly" },  { emoji: "🐢", name: "Turtle" },
  { emoji: "🐙", name: "Octopus" },    { emoji: "🐬", name: "Dolphin" },
  { emoji: "🐳", name: "Whale" },      { emoji: "🦈", name: "Shark" },
  { emoji: "🐠", name: "Fish" },       { emoji: "🦀", name: "Crab" },
  { emoji: "🐌", name: "Snail" },      { emoji: "🐞", name: "Ladybug" },
  { emoji: "🦄", name: "Unicorn" },    { emoji: "🐴", name: "Horse" },
  { emoji: "🐮", name: "Cow" },        { emoji: "🐷", name: "Pig" },
  { emoji: "🐰", name: "Rabbit" },     { emoji: "🐨", name: "Koala" },
  { emoji: "🐻", name: "Bear" },       { emoji: "🦒", name: "Giraffe" },
  { emoji: "🐘", name: "Elephant" },   { emoji: "🦓", name: "Zebra" },
  { emoji: "🦔", name: "Hedgehog" },   { emoji: "🦇", name: "Bat" },
  { emoji: "🦜", name: "Parrot" },     { emoji: "🦚", name: "Peacock" },
  { emoji: "🍎", name: "Apple" },      { emoji: "🍌", name: "Banana" },
  { emoji: "🍓", name: "Strawberry" }, { emoji: "🍒", name: "Cherry" },
  { emoji: "🍇", name: "Grapes" },     { emoji: "🍉", name: "Watermelon" },
  { emoji: "🍑", name: "Peach" },      { emoji: "🍍", name: "Pineapple" },
  { emoji: "🥝", name: "Kiwi" },       { emoji: "🥥", name: "Coconut" },
  { emoji: "🌽", name: "Corn" },       { emoji: "🥕", name: "Carrot" },
  { emoji: "🍄", name: "Mushroom" },   { emoji: "🌶️", name: "Pepper" },
  { emoji: "🍕", name: "Pizza" },      { emoji: "🍔", name: "Burger" },
  { emoji: "🌮", name: "Taco" },       { emoji: "🍩", name: "Donut" },
  { emoji: "🍪", name: "Cookie" },     { emoji: "🎂", name: "Cake" },
  { emoji: "🍦", name: "Ice cream" },  { emoji: "🍿", name: "Popcorn" },
  { emoji: "☕", name: "Coffee" },     { emoji: "🍵", name: "Tea" },
  { emoji: "⚽", name: "Soccer" },     { emoji: "🏀", name: "Basketball" },
  { emoji: "🏈", name: "Football" },   { emoji: "🎾", name: "Tennis" },
  { emoji: "🎱", name: "8-ball" },     { emoji: "🎯", name: "Target" },
  { emoji: "🎲", name: "Dice" },       { emoji: "🎮", name: "Game" },
  { emoji: "🎸", name: "Guitar" },     { emoji: "🎺", name: "Trumpet" },
  { emoji: "🎻", name: "Violin" },     { emoji: "🥁", name: "Drum" },
  { emoji: "🎹", name: "Piano" },      { emoji: "🎤", name: "Mic" },
  { emoji: "🎧", name: "Headphones" }, { emoji: "🚗", name: "Car" },
  { emoji: "🚀", name: "Rocket" },     { emoji: "✈️", name: "Plane" },
  { emoji: "🚲", name: "Bike" },       { emoji: "⛵", name: "Sailboat" },
  { emoji: "🚁", name: "Helicopter" }, { emoji: "🚂", name: "Train" },
  { emoji: "⚓", name: "Anchor" },     { emoji: "🪂", name: "Parachute" },
  { emoji: "🌙", name: "Moon" },       { emoji: "⭐", name: "Star" },
  { emoji: "☀️", name: "Sun" },        { emoji: "⚡", name: "Lightning" },
  { emoji: "🔥", name: "Fire" },       { emoji: "❄️", name: "Snowflake" },
  { emoji: "🌈", name: "Rainbow" },    { emoji: "🌻", name: "Sunflower" },
  { emoji: "🌹", name: "Rose" },       { emoji: "🌵", name: "Cactus" },
  { emoji: "🌲", name: "Tree" },       { emoji: "🍀", name: "Clover" },
  { emoji: "💎", name: "Diamond" },    { emoji: "🔑", name: "Key" },
  { emoji: "🔔", name: "Bell" },       { emoji: "🎁", name: "Gift" },
];

export function sasToGlyphs(sas: string | number): SasGlyph[] {
  let v = typeof sas === "number" ? sas : parseInt(sas, 10);
  if (!Number.isFinite(v) || v < 0) v = 0;
  v %= 1_000_000;
  const a = Math.floor(v / 10_000) % 100;
  const b = Math.floor(v / 100) % 100;
  const c = v % 100;
  return [SAS_EMOJI[a], SAS_EMOJI[b], SAS_EMOJI[c]];
}

File: android/app/src/main/java/com/vortex/a3/core/pairing/SasEmoji.kt
package com.vortex.a3.core.pairing

object SasEmoji {
    data class Glyph(val emoji: String, val name: String)

    val TABLE: List<Glyph> = listOf(
        Glyph("🦊", "Fox"), Glyph("🐼", "Panda"), Glyph("🦁", "Lion"), Glyph("🐯", "Tiger"),
        Glyph("🐶", "Dog"), Glyph("🐱", "Cat"), Glyph("🐵", "Monkey"), Glyph("🐸", "Frog"),
        Glyph("🐧", "Penguin"), Glyph("🦉", "Owl"), Glyph("🦅", "Eagle"), Glyph("🐝", "Bee"),
        Glyph("🦋", "Butterfly"), Glyph("🐢", "Turtle"), Glyph("🐙", "Octopus"), Glyph("🐬", "Dolphin"),
        Glyph("🐳", "Whale"), Glyph("🦈", "Shark"), Glyph("🐠", "Fish"), Glyph("🦀", "Crab"),
        Glyph("🐌", "Snail"), Glyph("🐞", "Ladybug"), Glyph("🦄", "Unicorn"), Glyph("🐴", "Horse"),
        Glyph("🐮", "Cow"), Glyph("🐷", "Pig"), Glyph("🐰", "Rabbit"), Glyph("🐨", "Koala"),
        Glyph("🐻", "Bear"), Glyph("🦒", "Giraffe"), Glyph("🐘", "Elephant"), Glyph("🦓", "Zebra"),
        Glyph("🦔", "Hedgehog"), Glyph("🦇", "Bat"), Glyph("🦜", "Parrot"), Glyph("🦚", "Peacock"),
        Glyph("🍎", "Apple"), Glyph("🍌", "Banana"), Glyph("🍓", "Strawberry"), Glyph("🍒", "Cherry"),
        Glyph("🍇", "Grapes"), Glyph("🍉", "Watermelon"), Glyph("🍑", "Peach"), Glyph("🍍", "Pineapple"),
        Glyph("🥝", "Kiwi"), Glyph("🥥", "Coconut"), Glyph("🌽", "Corn"), Glyph("🥕", "Carrot"),
        Glyph("🍄", "Mushroom"), Glyph("🌶️", "Pepper"), Glyph("🍕", "Pizza"), Glyph("🍔", "Burger"),
        Glyph("🌮", "Taco"), Glyph("🍩", "Donut"), Glyph("🍪", "Cookie"), Glyph("🎂", "Cake"),
        Glyph("🍦", "Ice cream"), Glyph("🍿", "Popcorn"), Glyph("☕", "Coffee"), Glyph("🍵", "Tea"),
        Glyph("⚽", "Soccer"), Glyph("🏀", "Basketball"), Glyph("🏈", "Football"), Glyph("🎾", "Tennis"),
        Glyph("🎱", "8-ball"), Glyph("🎯", "Target"), Glyph("🎲", "Dice"), Glyph("🎮", "Game"),
        Glyph("🎸", "Guitar"), Glyph("🎺", "Trumpet"), Glyph("🎻", "Violin"), Glyph("🥁", "Drum"),
        Glyph("🎹", "Piano"), Glyph("🎤", "Mic"), Glyph("🎧", "Headphones"), Glyph("🚗", "Car"),
        Glyph("🚀", "Rocket"), Glyph("✈️", "Plane"), Glyph("🚲", "Bike"), Glyph("⛵", "Sailboat"),
        Glyph("🚁", "Helicopter"), Glyph("🚂", "Train"), Glyph("⚓", "Anchor"), Glyph("🪂", "Parachute"),
        Glyph("🌙", "Moon"), Glyph("⭐", "Star"), Glyph("☀️", "Sun"), Glyph("⚡", "Lightning"),
        Glyph("🔥", "Fire"), Glyph("❄️", "Snowflake"), Glyph("🌈", "Rainbow"), Glyph("🌻", "Sunflower"),
        Glyph("🌹", "Rose"), Glyph("🌵", "Cactus"), Glyph("🌲", "Tree"), Glyph("🍀", "Clover"),
        Glyph("💎", "Diamond"), Glyph("🔑", "Key"), Glyph("🔔", "Bell"), Glyph("🎁", "Gift"),
    )

    fun glyphs(sas: String): List<Glyph> {
        val v = (sas.toIntOrNull() ?: 0).let { if (it < 0) 0 else it } % 1_000_000
        val a = (v / 10_000) % 100
        val b = (v / 100) % 100
        val c = v % 100
        return listOf(TABLE[a], TABLE[b], TABLE[c])
    }
}
